crates/shared/src/order_validation.rs

```diff
@@ -690,16 +667,42 @@ impl OrderValidating for OrderValidator {
                         buy: quote.buy_amount,
                         fee: quote.fee_amount,
                     },
-                ) =>
-            {
-                tracing::debug!(%uid, ?owner, ?class, "order being flagged as outside market price");
-                OrderClass::Liquidity
+                ) {
+                    tracing::debug!(%uid, ?owner, ?class, "order being flagged as outside market price");
+                    (OrderClass::Limit, Some(quote))
+                } else {
+                    (class, Some(quote))
+                }
+            }
+            OrderClass::Limit => {
+                let quote = get_quote_and_check_fee(
+                    &*self.quoter,
+                    &quote_parameters,
+                    order.quote_id,
+                    None,
+                    self.market_orders_deprecation_date,
+                )
+                .await?;
+                // If the order is not "In-Market", check for the limit orders
+                if is_order_outside_market_price(
+                    &Amounts {
+                        sell: data.sell_amount,
+                        buy: data.buy_amount,
+                        fee: data.fee_amount,
+                    },
+                    &Amounts {
+                        sell: quote.sell_amount,
+                        buy: quote.buy_amount,
+                        fee: quote.fee_amount,
+                    },
+                ) {
+                    self.check_max_limit_orders(owner).await?;
+                }
+                (class, Some(quote))
             }
-            (_, _) => class,
+            OrderClass::Liquidity => (OrderClass::Limit, None),
```

Comment:
Should we also `check_max_limit_orders` here? Also, I'm not sure if this code path gets hit in practice, but I believe we should also fetch a quote here as we are starting to make assumptions that each order is associated with a quote (and we want to trade liquidity orders like normal limit orders).

Reply:
@fleupold you are definitely right! I added the check. The `match` is a little bit ugly, but it is going to be totally refactor in my follow up task https://github.com/cowprotocol/services/issues/2469